content/articles/memoization-and-dp-in-python-beginners-guide/index.md

```diff
@@ -0,0 +1,122 @@
+###A Beginner's guide to memoization and Dynamic Programming in Python
+
+as we saw in [this]('https://www.section.io/engineering-education/introduction-to-dynamic-programming/) article, Dynamic programming is the algorithm technique where we break down a problem into smaller, simpler subproblems then find the solutions to these smaller subproblems and use these solutions to find an optimal solution to the original problem.
+In this article, we are going to talk about a dynamic programming technique called memoization and we are going to use it to solve an optimization problem.
+
+###Table of contents
+[Prerequisites](#prerequisites)
+
+[Memoization](#memoization)
+
+[The 0-1 knapsack problem](#knapsack)
+
+[Python Solution to the 0-1 knapsack problem](#implementation)
+
+[How to apply memoization to the solution](#solution)
+
+###prerequisites<a name='prerequisites'></a>
+1. A basic understanding of data structures and algorithms
+2. Basic knowledge of Python
+3. A basic understanding of Recursion
+### what is memoization? <a name='memoization'></a>
+memoization is the process of saving the results of a function to a cache in order to improve its performance.
+It is the technique where we ensure that the function does not run more than once for the same inputs; instead, we return a cached result to reduce the running time.
+It is similar to tabulation except that while tabulation uses the bottom-up approach, memoization uses a top-down approach. This means that when approaching a problem with tabulation, we save the results starting with smaller, lower values going up whereas in memoization we start from the larger values going down.
+In the fibonacci example given in the tutorial linked above we can see that we started from values 2 up towards the value passed on to the function.
+```python
+for i in range(2, n):
+   term = array[i - 1] + array[i - 2]
+   array.append(term)
+   # return the nth term of the array
+return array[n - 1]
+```
+The memoized version of this is going to look like this:
+```python
+def mem_fib(n, array):
+    '''
+    n is the fib number,
+    array is an array of n+1 numbers
+    '''
+    if n<=1:
+        array[n]=n
+    if array[n] is None:
+        array[n] = mem_fib(n-1, array)+mem_fib(n-2, array)
+    return array[n]
+```
+In the memoized version you can see that we started from the values n-1 and n-2 going downwards.
+Now that you know what memoization is, let's use it to solve a common programming problem.
```

Comment:
Don't just blindly state facts or copy stuff from other articles.

Instead walk the reader through the content. Explain them step-by-step. Explain how it works. Why we memoize it? What if we do not? Stack trace the program for sample input